src/PubSub/ResourceNameTrait.php

```diff
@@ -0,0 +1,144 @@
+<?php
+/**
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\PubSub;
+
+use InvalidArgumentException;
+
+trait ResourceNameTrait
+{
+    /**
+     * @var array
+     */
+    private $templates = [
+        'project' => 'projects/%1$s',
+        'topic' => 'projects/%2$s/topics/%1$s',
+        'subscription' => 'projects/%2$s/subscriptions/%1$s'
+    ];
+
+    /**
+     * @var array
+     */
+    private $regexes = [
+        'project' => '/^projects\/([^\/]*)$/',
+        'topic' => '/projects\/[^\/]*\/topics\/(.*)/',
```

Comment:
The topic and subscription has a well-defined rule:

https://cloud.google.com/pubsub/overview

```
The resource-name must start with a letter, and contain only letters ([A-Za-z]), numbers ([0-9], dashes (-), underscores (_), periods (.), tildes (~), plus (+) or percent signs (%). It must be between 3 and 255 characters in length, and cannot begin with the string goog.
```

Is it worthwhile to check it earlier than later?